waiter/src/waiter/state.clj

```diff
@@ -1303,3 +1308,114 @@
+  (let [{:strs [min-instances]} (service-id->service-description-fn service-id)
+        ctxs (repeatedly
+               min-instances
+               (fn [] {:schedule [(timers/start (metrics/waiter-timer "service-init-overhead" "schedule-time"))
+                                  (timers/start (metrics/service-timer service-id "service-init-overhead" "schedule-time"))]
+                       :total [(timers/start (metrics/waiter-timer "service-init-overhead" "init-to-healthy-time"))
+                               (timers/start (metrics/service-timer service-id "service-init-overhead" "init-to-healthy-time"))]}))]
+    (map->InstanceTracker
+      {:service-id service-id
+       :known-instance-ids #{}
+       :scheduling-instance-timer-contexts (vec ctxs)
+       :starting-instance-trackers []})))
+
+(defn update-instance-trackers
+  "Track metrics on app instances, specifically schedule and startup times."
+  [service-instance-trackers new-service-ids removed-service-ids service-id->healthy-instances service-id->unhealthy-instances service-id->service-description-fn]
+  (->> service-instance-trackers
+       (remove (comp removed-service-ids :service-id))
+       (concat (mapv #(new-tracker-val % service-id->service-description-fn) new-service-ids))
+       (mapv (fn [{:keys [service-id known-instance-ids scheduling-instance-timer-contexts starting-instance-trackers]}]
+               (let [healthy-instance-ids (->> service-id
+                                               (get service-id->healthy-instances)
+                                               (map :id)
+                                               set)
+                     known-instance-ids' (->> service-id
+                                              (get service-id->unhealthy-instances)
+                                              (map :id)
+                                              (into healthy-instance-ids))
+                     new-instance-ids (set/difference known-instance-ids' known-instance-ids)
+                     ;; Pair up all of the newly-scheduled instances with their timers.
+                     ;; NOTE: Using split-at allows us to drop the new-instances if we're somehow missing the corresponding timers
+                     [paired-timers scheduling-instance-timer-contexts'] (split-at (count new-instance-ids) scheduling-instance-timer-contexts)
+                     starting-instance-trackers' (->>
+                                                   ;; Start timers for startup time of newly-scheduled instances.
+                                                   ;; NOTE: Since Waiter only scales down by killing known instances, waiting-to-be-scheduled instances are not removed (unlike starting instances).
+                                                   (map (fn [instance-id {schedule-timers :schedule total-timers :total}]
+                                                          (doseq [timer schedule-timers] (timers/stop timer))
```

Comment:
Nitpick: prefer newline for `doseq` body

Reply:
`timer` should be named `timer-context`

Reply:
Both done.